sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java

```diff
@@ -58,6 +58,58 @@
 
 /**
  * A {@link ChannelPool} implementation that enforces a maximum number of concurrent direct TCP Cosmos connections.
+ *
+ * RntbdClientChannelPool: Actors
+ * 	- acquire (RntbdServiceEndpoint): acquire a channel to use
+ * 	- release (RntbdServiceEndpoint): channel usage is complete and returning it back to pool
+ * 	- Channel.closeChannel() Future: Event handling notifying the channel termination to refresh bookkeeping
+ * 	- acquisitionTimeoutTimer: channel acquisition time-out handler
+ * 	- monitoring (through RntbdServiceEndpoint): get monitoring metrics
+ *
+ * 	Behaviors/Expectations:
+ * 	    - Bounds:
+ * 	        - max requests in-flight per channelPool: MAX_CHANNELS_PER_ENDPOINT * MAX_REQUESTS_ENDPOINT (NOT A GUARANTEE)
+ * 	        - AvailableChannels.size() + AcquiredChannels.size() <= MAX_CHANNELS_PER_ENDPOINT
+ * 	        - PendingAcquisition queue default-size: Max(10_000, MAX_CHANNELS_PER_ENDPOINT * MAX_REQUESTS_ENDPOINT)
+ * 	        - ChannelPool executor event-loop task queue length: MAX_CHANNELS_PER_ENDPOINT * MAX_REQUESTS_ENDPOINT + newInFlightAcquisitions (not yet in pendingAcquisitionQueue)
+ * 	            - newInFlightAcquisitions: is expected to very very short. Hard-bound to ADMINSSON_CONTROL (upstream in RntbdServiceEndpoint)
+ * 	    - NewChannel vs ReUseChannel:
+ * 	        - NewChannels are serially created (reasonable current state, possible future change, upstream please DON'T TAKE any dependency)
+ * 	        - Will re-use an existing channel when possible (with MAX_REQUESTS_ENDPOINT attempt not GUARANTEED)
+ * 	        - Channel usage fairness: fairness is attempted but not guaranteed
+ * 	            - When loadFactor is > 90%, fairness is attempted by selecting Channel with less concurrency
+ * 	            - Otherwise no guarantees on fairness per channel with-in bounds of MAX_REQUESTS_ENDPOINT. I.e. some channel might have high request concurrency compared to others
+ * 	    - Channel serving guarantees:
+ * 	        - Ordered delivery is not guaranteed (by-design)
+ * 	        - Fairness is attempted but not a guarantee
+ * 	        - [UNRELATED TO CHANNEL-POOL] [CURRENT DESIGN]: RntbdServiceEndpoint.write releases Channel before its usage -> acquisition order and channel user order might differ.
+ * 	    - AcquisitionTimeout: if not can't be served in an expected time, fails gracefully
+ * 	    - Metrics: are approximations and might be in-consistent(by-design) as well
+ * 	    - EventLoop
+ * 	        - ChannelPool executor might be shared across ChannelPools or Channel
+ *
+ * 	Design Notes:
+ * 	    - channelPool.eventLoop{@Link executor}: (executes on a single & same thread, serially)
+ * 	        - Each channelPool gets an EventLoop (selection is round-robin)
+ * 	        - Schedule only when it can be served immediately
+ * 	        - Updates and reads that depend on "strong consistency" - like whether to create a new connection or not.
```

Comment:
`strong consistency` does this refer to cosmos db consistency?

The reason the access to internal fields need to happen on event-loop is to prevent race condition. I cannot see how "cosmos strong consistency" is relevant here.

Reply:
Closed offline